### PR TITLE
Paper: Funix 

### PR DESCRIPTION
**Pull request by forrestbao**
If you are creating this PR in order to submit a draft of your paper, please name your PR with `Paper: <title>`. An editor will then add a `paper` label and GitHub Actions will be run to check and build your paper.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#preview-your-paper) for more information.

_Editor_: Chris Calloway @cbcunc

_Reviewers_: 
- Chong Shen Ng @chongshenng
- Conrad Koziol @cpk26

### CODE REVIEW

**Comment by stevejpurves** [comment]
Hi @forrestbao - looks like you have made changes to `.gitignore` outside of the paper folder, could you please revert that change (adding a `.gitignore` in your paper folder if you need to), then the checks will pass.

**Comment by cbcunc** [comment]
Review reminders sent to @chongshenng and @cpk26

**Comment by chongshenng** [comment]
Hi @forrestbao 👋, I'm Chong Shen one of the reviewers. I'll add my comments to your PR shortly.

File: papers/funix/main.md [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+When it comes to GUI app development, there is a trade-off between simplicity and versatility.
+JavaScript/TypeScript-based web frontend frameworks like React, Angular, and Vue.js are versatile.
+But that versatility is beyond the reach of most scientists and engineers, except frontend/full-stack engineers, and is usually overkilling for most scientific and engineering apps.
+
+As machine learning researchers ourselves, we notice that a great (if not overwhemling) amount of scientific apps have the following two features:
+
+1. the underlying logic is a straightforward input-output process -- thus complex interactivity, such as updating the input options based on existing user input, is not needed;
+2. the app is not the goal but a necessary step to the goal -- thus it is not worthy to spend time on building the app.
+
+Existing Python-based solutions such as Streamlit or Grdio, do a great job for the first feature above but are still too complicated for the second one, requiring a developer to read their documentations and add some code before an app can be fired up.
```
Comment (chongshenng) [inline papers/funix/main.md]: Tiny typo here: Grdio -> Gradio

File: papers/funix/main.md [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+
+A Dall-E app generated by Funix by simply wraping OpenAI's image generation API with a `str`-to-`Image` function. Source code in [](#code_dalle). 
+
+```
+
+## Funix's default mapping from Python types to React components
+
+Like CSS, Funix controls the GUI appearance based on the types of arguments and returns of a function.
+Under the hood, Funix transcompiles the signature of a function into React code.
+Currently, Funix depends on the type hint for every argument or return of a function. In the future, it will support type inference or tracing.
```
Comment (chongshenng) [inline papers/funix/main.md:149]: > In the future, it will support type inference or tracing.

❤️

File: papers/funix/main.md [hunk 3]
```diff
@@ -0,0 +1,792 @@
...
+
+By default, Funix maps the following basic Python types to the following MUI components:
```
Comment (chongshenng) [inline papers/funix/main.md]: I'm assuming the users don't have to explicitly remember this mapping, if you push the boundaries of simplicity? Could you please comment on this?

File: papers/funix/main.md [hunk 4]
```diff
@@ -0,0 +1,792 @@
...
+### Output layout in `print` and `return`
+
+In Funix, by default, the return of a function becomes the content of the output panel.
+A user can control the layout of the output panel returning strings, including and especially f-strings, in the Markdown and HTML syntaxes.
+Markdown and HTML strings must be explicitly specified of the types `IPython.display.Markdown` and `IPython.display.HTML`, respectively. Otherwise, the raw strings will be displayed. Because Python supports multiple returns, you can mix Markdown and HTML strings in the return statement.
+
+Quite often we need to print out some information before a function reaches its returns.
+`print` is a Python built-in function that is frequently used for this purpose.
+Funix extends this convenience by redirecting the output of `print` to the output panel of an app. The printout strings in Markdown or HTML syntax and will be automatically rendered after syntax detection.
+To avoid conflicting with the default behavior of printing to `stdout`, printing to the web needs to be explicitly turned on by a decorator.
```
Comment (chongshenng) [inline papers/funix/main.md]: Pretty cool feature. I was wondering how does `funix` handle `stderr`? For instance when using `logging.Logger` objects. It might be out of scope of this paper, so I'll leave it up to the authors to decide whether or not to mention this, or if it's something further down the line for development.

**Review by chongshenng — CHANGES_REQUESTED** [review]
The authors present Funix, a supercharged hyper-simple way of converting Python native scripts/functions into a GUI _without_ needing to know about front-end or app development. Funix is an early stage project, but it shows tremendous promise and utility for scientific users. I am looking forward to this paper in its full form. 

In terms of comments, I only have very minor ones (please refer to the inline comments0. The writing is clear and the contents are informative. In fact, reading this paper is almost like reading the documentation of Funix (which is interesting because the authors wrote in their Motivation: "... all existing solutions require the developer to read their documentations before being able to create a GUI." 😉 ). 

To summarize, I only have minor comments and happy for the authors to amend them accordingly, where they think are appropriate.

File: papers/funix/main.md (hunk 4 above)
Comment (Konoyoru435) [reply papers/funix/main.md]: Currently, Funix wraps the output in `stderr` in `<code>` to distinguish it from `stdout`, so perhaps in the future Funix will think about handling `stderr` and `stdout` in a better way.

**Comment by cpk26** [comment]
This paper introduces the early-stage Python project Funix. The aim of Funix is to automatically generate a web-based GUI for Python programs based on the types of function parameters and return values. It differentiates itself from existing frameworks by not requiring users to explicitly select the GUI elements used. Rather, Funix automatically selects elements based on types such as text boxes for string parameters or checkboxes for Boolean parameters. The correspondence between types and elements can be modified via themes. The motivation for Funix are the observations: that scientists and engineers are not necessarily experts in front-end development, that numerous use-cases exist for 'disposable' frontends where the idea is to provide a minimum-viable product or proof-of-concept, and that there exists relatively straight forward relationships between input/output elements and parameter types. The underlying mechanics of the project are implemented by transpiling Python into React.


Overall my impression of the project and paper are positive. It's well motivated in its discussion of simplifying GUI development for non-frontend developers, the mechanics of using Funix are described using examples, and there are three demos showcased at the end. As mentioned at the end of the paper, it has parallels to frameworks like Fire which automate generating a command line interface, which are quite useful. This paper makes good use of code examples and corresponding figures. The writing however, could use another pass at polishing. This includes fixes to the grammar, clarifying certain non-obvious points, and removing unnecessary opinions. See individual comments for more details.

While I did not run all the examples, getting the hello world example running was frictionless.

File: papers/funix/main.md [hunk 5]
```diff
@@ -0,0 +1,792 @@
+---
+# Ensure that this title is the same as the one in `myst.yml`
+title: Funix - vivifying Python functions into GUI apps
+abstract: |
+  The raise of machine learning (ML) and artificial intelligence (AI), especially the generative AI (GenAI), has brought up the need for wrapping models or algorithms into GUI apps. For example, a large language model (LLM) can be accessed through a string-to-string GUI app with a textbox as the primary input.
```
Comment (cpk26) [inline papers/funix/main.md]: 'raise' -> 'rise' (plus in other locations)

File: papers/funix/main.md [hunk 6]
```diff
@@ -0,0 +1,792 @@
...
+  Most of existing solutions require developers to manually create widgets and link them to arguments/returns of a function individually. This low-level process is laborious and usually intrusive. Funix takes a CSS-like approach by automatically picking widgets based on the types of the arguments and returns of a function according to the type-to-widget mapping defined in a theme, e.g., `bool` to a checkbox. As a result, an existing Python function can be turned into a GUI app without any code change. As a transcompiler, Funix allows such type-to-widget mapping to be defined between any Python type  and any React component and its `props`, liberating Python developers to the frontend world without needing to know JavaScript/TypeScript. Funix further uses the information in Docstrings, which are common in Python development, to control the appearance of the GUI.
```
Comment (cpk26) [inline papers/funix/main.md]: The connection to CSS is unclear to me. Perhaps expand a little? This reference is used a few times through the paper.

File: papers/funix/main.md [hunk 7]
```diff
@@ -0,0 +1,792 @@
...
+Most of them follow the conventional GUI programming philosophy that a developer needs to manually pick widgets from a widget library and associate them with the arguments and returns of an underlying function, which is usually called the "callback function."
+
+This approach has several drawbacks. **First**, it is a repetitive, manual process. A developer has to manually align and keep aligning the signature of the callback function with the GUI code. The dependency of between the two means that one can be derived from the other and the process could have be automated. **Second**, the developer is bounded by the widgets provided by a GUI library. **Third**, they do not leverage the features of the Python language itself to automate the process. For example, most existing solutions require developers to manually specify the labels of widgets while such information is usually in the `params` or `args` sections in the docstrings of a function, which are very common in Python development. **Last** but not least, all existing solutions require the developer to read their documentations before being able to create a GUI, even one as simple as for saying "hello, world!”
+
```
Comment (cpk26) [inline papers/funix/main.md:18]: Very ambitious to try to eliminate reading the step of reading the docs!

File: papers/funix/main.md [hunk 8]
```diff
@@ -0,0 +1,792 @@
...
+As a result, scientific developers, such as geophysicists, neurobiologists, or machine learning engineers, whose jobs are not building apps, are not able to quickly fire up apps to present their models, algorithms, or discoveries to the world.
+Every minute they spent on learning and building apps in existing solutions is a bad investment of their time because it does not advance their domain expertise or career goals.
```
Comment (cpk26) [inline papers/funix/main.md]: I think there's varied opinions on whether this is a 'bad investment', or entirely separate from domain expertise and career goals. Bridging the gap between science and practice is important. This sentence might be better removed, as the motivation to help scientists do it faster is sufficient in the previous sentence is sufficient.

File: papers/funix/main.md [hunk 9]
```diff
@@ -0,0 +1,792 @@
...
+Every minute they spent on learning and building apps in existing solutions is not advancing their domain expertise or career goals. -->
+
+Therefore, [Funix](http://funix.io) was born to automatically launch apps from existing functions. We notice that the choice of a widget has a weak correlation with the type of the function I/O it is associated with. For example, a checkbox is unsuitable for any type other than Boolean. Therefore, Funix takes a CSS-like approach to automatically pick widgets based on the types of the arguments/returns of a function. For example, in the default theme of Funix, the Python native types `str`, `bool`, and `Literal`, respectively, are mapped into an input box, a checkbox, and a set of radio buttons in the MUI library, while the common scientific types `pandas.DataFrame` and `matplotlib.figure.Figure` are mapped to tables (MUI's `DataGrid`) and charts (in `mpld3`). A variable's type can be specified in type hints, which are a good and common practice in Python development, or inferred, which is to be supported in the future by Funix.
```
Comment (cpk26) [inline papers/funix/main.md]: MUI library -- This should be introduced, or at least linked to?

File: papers/funix/main.md [hunk 10]
```diff
@@ -0,0 +1,792 @@
...
+
+Besides types, Funix also makes use other features in the Python language or ecosystem to further automate app building. Docstrings are dominantly common in Python. Funix uses the information in Docstrings to control the UI appearance. For example, the annotation of each argument in the `Args` section in Google-style Docstrings or `Parameters` section in Numpy-style Docstrings can become the label/tooltip to explain the meaning of the argument to the app user. 
+Funix also give some keywords/types in Python new meanings in the context of app building, e.g., `global` for states and sessions, `yield` for streaming, and each class becomes a multi-page app where pages share data via the `self` variable. 
+
+Funix is not only a GUI generator. It is a transcompiler that generates both the backend and the frontend of an app. The two ends communicate with each other in websocket. For the backend, Funix wraps a Python function into a Flask app. Thus the Flask app can also be access programmatically. For the frontend, it generates React code. In Funix, a Python type can be bound to any React component and its `props` via a theme.
```
Comment (cpk26) [inline papers/funix/main.md]: May be worth introducing the concept of a transpiler with a sentence or two as you are targeting scientists + engineers, not necessarily computer scientists.

File: papers/funix/main.md [hunk 11]
```diff
@@ -0,0 +1,792 @@
...
+
+Funix is not only a GUI generator. It is a transcompiler that generates both the backend and the frontend of an app. The two ends communicate with each other in websocket. For the backend, Funix wraps a Python function into a Flask app. Thus the Flask app can also be access programmatically. For the frontend, it generates React code. In Funix, a Python type can be bound to any React component and its `props` via a theme. 
+Therefore, Funix allows Python developers to tap into the frontend world without JavaScript/TypeScript knowledge.
+Consequently, Python becomes a surface language for web development. 
+
+Funix can be really useful in building what we call "disposable apps" -- apps that are built for short-term/non-permanent purposes with a small group of audiences, such as demoing PoCs, trying out an AI model, collecting human labeling of AI training data, exhibiting the capabilities of an API, etc.
+Such disposable apps maybe recurrently needed and 
+Funix allows them to be launched rapidly again and again. 
+
+In summary, Funix has the following cool features to make building apps effortless (if not lazy or cheating):
```
Comment (cpk26) [inline papers/funix/main.md]: "(if not lazy or cheating)" -- This sentence can be removed, as the language seems out of place. I think the preceding statement communicates the idea well.

File: papers/funix/main.md [hunk 12]
```diff
@@ -0,0 +1,792 @@
...
+
+    response = client.images.generate(Prompt)
+
+    return response.data[0].url
+```
+
+```{figure} dalle.png
+:label: fig_dalle
+
+A Dall-E app generated by Funix by simply wraping OpenAI's image generation API with a `str`-to-`Image` function. Source code in [](#code_dalle).
```
Comment (cpk26) [inline papers/funix/main.md]: "wraping" -> wrapping

File: papers/funix/main.md [hunk 13]
```diff
@@ -0,0 +1,792 @@
...
+
+## Funix's default mapping from Python types to React components
+
+Like CSS, Funix controls the GUI appearance based on the types of arguments and returns of a function.
+Under the hood, Funix transcompiles the signature of a function into React code.
+Currently, Funix depends on the type hint for every argument or return of a function. In the future, it will support type inference or tracing.
+
+By default, Funix maps the following basic Python types to the following MUI components:
+
+| Python type     | Input or Output | MUI component or HTML                                                                                                                                                                 |
```
Comment (cpk26) [inline papers/funix/main.md]: Does this table cover how the cat is displayed in the previous code example?

File: papers/funix/main.md [hunk 14]
```diff
@@ -0,0 +1,792 @@
...
+```{figure} table_and_chart.png
+:label: fig_table_and_chart
+
+An app generated from [](#code_table_and_chart) by Funix. The input panel contains a table and the output panel contains a chart. Both the table and the chart are interactive/editable. As far as we know, no other Python-based solutions supports editable tables as inputs.
+```
+
+## Defining new type-to-widget mappings
+
+:::{note}
+Introducing a new type-to-widget mapping or modifying an existing one should not be the job of most Funix users. Only advanced users or the UI guys should be involved. This is like most scientists who write papers in LaTeX do not develop the LaTeX classes or macros but just use them.
```
Comment (cpk26) [inline papers/funix/main.md]: "Only advanced users or the UI guys should be involved." -- This sentence can be removed, or should be reworded with more precise language to maintain a professional tone. Perhaps something along the lines of: "This can be limited to advanced users or user interface specialists."?

File: papers/funix/main.md [hunk 15]
```diff
@@ -0,0 +1,792 @@
...
+```
+
+### Streaming based on `yield`
+
+In this GenAI era, streaming is becoming a common way for returning lengthy text output from an AI model.
+Instead of inventing something new to support streaming, Funix repurposes the `yield` keyword in Python to stream the output of a function. The rationale is that `return` and `yield` are highly similar and `return` has already been used to display the final output in a Funix-powered app.
+
+```{code} python
+:label: code_stream
+:caption: Stremaing using yield in Funix. The corresponding GUI app is shown in [](#fig_stream).
```
Comment (cpk26) [inline papers/funix/main.md]: Stremaing  -> streaming

File: papers/funix/main.md (hunk 1 above)
Comment (forrestbao) [reply papers/funix/main.md]: @chongshenng  Thanks very much. We will go though the paper a couple more times to carefully polish up the writing.

File: papers/funix/main.md (hunk 3 above)
Comment (forrestbao) [reply papers/funix/main.md]: Correct, users don't have to remember this mapping. The information here is just to document the correspondence between types and widgets in case some readers are curious.

File: papers/funix/main.md (hunk 5 above)
Comment (forrestbao) [reply papers/funix/main.md]: thanks for catching the error up. We will go through the paper a couple of times to polish the writing.

File: papers/funix/main.md (hunk 6 above)
Comment (forrestbao) [reply papers/funix/main.md]: Sure. As an abstraction, CSS manages the styles centrally (in one places, the `*.css` files) and separates the content from the style definitions. Changing the style for one class/id will update the looks of all tags of the class/id. Funix is like CSS in this sense. Widgets are grouped by their corresponding types. Changing the type-to-widget mapping in the theme will update the looks of all widgets of the involved type. Let me know if you have further questions.

File: papers/funix/main.md (hunk 7 above)
Comment (forrestbao) [reply papers/funix/main.md:18]: We simply want developer to be able to fire up something quickly. Of course, a developer needs to read docs for customization.

File: papers/funix/main.md (hunk 8 above)
Comment (forrestbao) [reply papers/funix/main.md]: sounds good. We will remove this sentence.

File: papers/funix/main.md (hunk 9 above)
Comment (forrestbao) [reply papers/funix/main.md]: We will link to it.

File: papers/funix/main.md (hunk 10 above)
Comment (forrestbao) [reply papers/funix/main.md]: sounds good. We will explain it.

File: papers/funix/main.md (hunk 11 above)
Comment (forrestbao) [reply papers/funix/main.md]: sounds good.

File: papers/funix/main.md (hunk 12 above)
Comment (forrestbao) [reply papers/funix/main.md]: We will polish the writing.

File: papers/funix/main.md (hunk 13 above)
Comment (forrestbao) [reply papers/funix/main.md]: In this table we only mention basic Python types. In the table below, we mention multimedia types, including the image type used to display the cat.

File: papers/funix/main.md (hunk 14 above)
Comment (forrestbao) [reply papers/funix/main.md]: We will remove it.

File: papers/funix/main.md (hunk 15 above)
Comment (forrestbao) [reply papers/funix/main.md]: Thanks. We will polish the writing

**Comment by forrestbao** [comment]
@chongshenng @cpk26 Thank you both for the positive reviews and detailed feedback. Glad you liked Funix. We will go over the paper a couple of times to polish the writing.

File: papers/funix/main.md (hunk 6 above)
Comment (cpk26) [reply papers/funix/main.md]: My impression is that this analogy isn't particularly helpful here. The reference to LaTeX themes later on was more useful, as its inline with your expectation that most folks take a stock theme and run with it. I'd recommend removing the reference to the CSS-like approach since it distracts from the rest of the sentence.

**Comment by chongshenng** [comment]
The authors have adequately addressed my comments in their revision.

**Comment by cpk26** [comment]
The revised manuscript addresses my comments as well.

**Comment by forrestbao** [comment]
Thank you very much @chongshenng and @cpk26.
